torch/nested/_internal/cached_tensor.py

```diff
@@ -0,0 +1,166 @@
+from contextlib import contextmanager
+from typing import *  # noqa: F403
+
+import torch
+from torch.utils import _pytree as pytree
+
+
+def _get_source_field(
+    metadata: Dict[str, torch.Tensor], source_fields: Tuple[str, ...]
+) -> str:
+    return next(k for k in source_fields if metadata.get(k) is not None)
+
+
+def _get_source(
+    metadata: Dict[str, torch.Tensor],
+    source_fields: Tuple[str, ...],
+    target_field: Optional[str],
+) -> torch.Tensor:
+    # Unless specified via target_field, CachedTensor's source is the first non-None entry in source_fields
+    if target_field is not None:
+        return metadata[target_field]
+    source_field = _get_source_field(metadata, source_fields)
+    return metadata[source_field]
+
+
+class CachedTensor(torch.Tensor):
+    @staticmethod
+    def __new__(
+        cls,
+        metadata: Dict[str, torch.Tensor],
+        source_fields: Tuple[str, ...],
+        extra_fields: Tuple[str, ...],
+        target_field: Optional[str] = None,
```

Comment:
is `target_field` used in NJT? didn't look like it to me but I could be wrong

Reply:
The original use case here was if a NJT's metadata holds both the lengths and offsets. If the user calls .offsets(), we'd want to return a metadata with target_field as offsets, and analgously for lengths.

I ended up updating .offsets()/._offsets to just return plain tensors, so this logic is no longer used by NJT today, but I imagine it would be if someone wanted a .offsets_with_cache() API.